### PR TITLE
iOS - Floating cursor

### LINKED ISSUE #88: iOS trackpad doesn't work
Hi there, holding down on the keyboard in iOS does not show the trackpad. This was recently added to Flutter's TextFields so perhaps some of the implementation can be stolen from that?

### PR DESCRIPTION
**Pull request by amantoux**
On iOS, ability to use virtual keyboard as a trackpad.
Closes #88 

https://user-images.githubusercontent.com/17318211/141678553-4d7e6499-61e9-48c0-86d8-f845984db84d.mov

### CODE REVIEW

File: packages/zefyr/lib/src/rendering/cursor_painter.dart [hunk 1]
```diff
@@ -99,3 +99,32 @@ class CursorPainter {
     return Offset(pixelPerfectOffsetX, pixelPerfectOffsetY);
   }
 }
+
```
Comment (amantoux) [inline packages/zefyr/lib/src/rendering/cursor_painter.dart:102]: Floating painter responsible to drawing the cursor when floating mode is activated

File: packages/zefyr/lib/src/rendering/editable_text_line.dart [hunk 2]
```diff
@@ -702,11 +718,16 @@ class RenderEditableTextLine extends RenderEditableBox {
   }
 
   void _paintCursor(PaintingContext context, Offset effectiveOffset) {
-    final position = TextPosition(
-      offset: selection.extentOffset - node.documentOffset,
-      affinity: selection.base.affinity,
-    );
-    _cursorPainter.paint(context.canvas, effectiveOffset, position);
+    var textPosition = _cursorController.isFloatingCursorActive
```
Comment (amantoux) [inline packages/zefyr/lib/src/rendering/editable_text_line.dart:721]: Paints regular cursor but also the background cursor when floating cursor is activated.
The background cursor shows the closest text position of the regular corresponding to the current floating cursor position should the latter be released

File: packages/zefyr/lib/src/rendering/editor.dart [hunk 3]
```diff
@@ -627,4 +658,127 @@ class RenderEditor extends RenderEditableContainerBox
     final boxParentData = targetChild.parentData as BoxParentData;
     return childLocalRect.shift(Offset(0, boxParentData.offset.dy));
   }
+
```
Comment (amantoux) [inline packages/zefyr/lib/src/rendering/editor.dart:661]: Floating cursor's painting is the responsibility of the render editor to ensure smoothness of the animation
The render editor now holds a reference to the cursor controller. It sets the closest text position `floatingCursorTextPosition` in the cursor controller. The latter will be read by `RenderEditableTextline` to paint the background cursor

File: packages/zefyr/lib/src/rendering/editor.dart [hunk 4]
```diff
@@ -627,4 +658,127 @@ class RenderEditor extends RenderEditableContainerBox
...
+    _resetFloatingCursorAnimationValue = resetLerpValue;
+    if (_floatingCursorOn) {
+      _floatingCursorTextPosition = textPosition;
+      final double? animationValue = _resetFloatingCursorAnimationValue;
+      final EdgeInsets sizeAdjustment = animationValue != null
+          ? EdgeInsets.lerp(
+              _kFloatingCaretSizeIncrease, EdgeInsets.zero, animationValue)!
+          : _kFloatingCaretSizeIncrease;
+      final child = childAtPosition(textPosition);
+      final caretPrototype =
```
Comment (amantoux) [inline packages/zefyr/lib/src/rendering/editor.dart:767]: `RenderEditableBox` requires `getCaretPrototype`

File: packages/zefyr/lib/src/rendering/editor.dart [hunk 5]
```diff
@@ -517,7 +537,7 @@ class RenderEditor extends RenderEditableContainerBox
   }
 
   // Call through to onSelectionChanged.
-  void _handleSelectionChange(
+  void handleSelectionChange(
```
Comment (pulyaevskiy) [inline packages/zefyr/lib/src/rendering/editor.dart:540]: Why making this method public?

File: packages/zefyr/lib/src/rendering/editor.dart [hunk 6]
```diff
@@ -627,4 +658,127 @@ class RenderEditor extends RenderEditableContainerBox
...
+
+    _previousOffset = rawCursorOffset;
+
+    return adjustedOffset;
+  }
+
+  @override
+  void setFloatingCursor(FloatingCursorDragState dragState,
+      Offset boundedOffset, TextPosition textPosition,
+      {double? resetLerpValue}) {
```
Comment (pulyaevskiy) [inline packages/zefyr/lib/src/rendering/editor.dart:748]: The original method definition in the abstract editor does not include `resetLerpValue`, is this intentional?
Would be useful to document what this argument does.

File: packages/zefyr/lib/src/widgets/editor_input_client_mixin.dart [hunk 7]
```diff
@@ -199,9 +201,115 @@ mixin RawEditorStateTextInputClientMixin on EditorState
...
+        if (_lastTextPosition != null && _lastBoundedOffset != null) {
+          floatingCursorResetController.value = 0.0;
+          floatingCursorResetController.animateTo(1.0,
+              duration: _floatingCursorResetTime, curve: Curves.decelerate);
+        }
+        break;
+    }
+  }
+
+  void onFloatingCursorResetTick() {
```
Comment (pulyaevskiy) [inline packages/zefyr/lib/src/widgets/editor_input_client_mixin.dart:286]: This would also benefit from a doc comment.

**Review by pulyaevskiy — CHANGES_REQUESTED** [review]
Thank you! This LGTM overall, just have a few comments around docs. PTAL

File: packages/zefyr/lib/src/rendering/editor.dart (hunk 5 above)
Comment (amantoux) [reply packages/zefyr/lib/src/rendering/editor.dart:540]: For a wrong reason, thanks

**Review by pulyaevskiy — APPROVED** [review]
LGTM, thank you!

**Pull request merged**